keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md

```diff
@@ -0,0 +1,891 @@
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+When a cluster has multiple apiservers at mixed versions (such as during an
+upgrade or downgrade), not every apiserver can serve every resource at every
+version.
+
+To fix this, we will add a filter to the handler chain in the aggregator which
+proxies clients to an apiserver that is capable of handling their request.
+
+## Motivation
+
+When an upgrade or downgrade is performed on a cluster, for some period of time
+the apiservers are at differing versions and are able to serve different sets of
+built-in resources (different groups, versions, and resources are all possible).
+
+In an ideal world, clients would be able to know about the entire set of
+available resources and perform operations on those resources without regard to
+which apiserver they happened to connect to. Currently this is not the case.
+
+Today, these things potentially differ:
+* Resources available somewhere in the cluster
+* Resources known by a client (i.e. read from discovery from some apiserver)
+* Resources that can be actuated by a client
+
+This can have serious consequences, such as namespace deletion being blocked
+incorrectly or objects being garbage collected mistakenly.
+
+### Goals
+
+* Ensure discovery reports the same set of resources everywhere (not just group
+  versions, as it does today)
+* Ensure that every resource in discovery can be accessed successfully
+* In the failure case (e.g. network not routable between apiservers), ensure
+  that unreachable resources are served 503 and not 404.
+
+### Non-Goals
+
+* Lock particular clients to particular versions
+
+## Proposal
+
+We will use the existing `StorageVersion` API to figure out which group, versions,
+and resources an apiserver can serve.
+
+
+API server change:
+* A new handler is added to the stack:
+
+  - If the request is for a group/version/resource the apiserver doesn't have
+    locally (we can use the StorageVersion API), it will proxy the request to
+    one of the apiservers that is listed in the object. If an apiserver fails
+    to respond, then we will return a 503 (there is a small
+    possibility of a race between the controller registering the apiserver
+    with the resources it can serve and receiving a request for a resource
+    that is not yet available on that apiserver).
+
+* Discovery merging.
+
+  - During upgrade or downgrade, it may be the case that no apiserver has a
+    complete list of available resources. To fix the problems mentioned, it's
+    necessary that discovery exactly matches the capability of the system. So,
+    we will use the storage version objects to reconstruct a merged discovery
```

Comment:
This is a new requirement for aggregated apiservers.  While reasonable (I think), finding a way to notify them of a mismatch between the StorageVersions they provide discovery they provide is important.  This may be covered below.

Reply:
We are not planning to change the handling of discovery requests for alpha. Merged discovery is being targeted for beta.